File: day5/typescript/day5.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

const doJumps = (change: (jump: number) => number) => (data: number[]) => {
    let offset = 0;
    let count = 0;
    for (; offset < data.length && offset >= 0; ++count) {
        const jump = data[offset];
        data[offset] += change(jump);
        offset += jump;
    }

    return count;
};

const part1 = doJumps(R.always(1));
const part2 = doJumps(jump => jump >= 3 ? -1 : 1);

async function day5() {
    const input = await promisify(readFile)('day5/input.txt', 'utf8');
    const jumps = R.map(parseInt, input.split('\r\n'));

    console.log(part1([...jumps]));
    console.log(part2([...jumps]));
}

day5();

File: day5/haskell/day5.hs
import Control.Lens
import Data.Array (Array, listArray)

type Tape = Array Int Int

parse :: String -> Tape
parse i = listArray (0, length offsets - 1) offsets
    where offsets = map read . lines $ i

jump :: (Int -> Int) -> Int -> Int -> Tape -> Int
jump f n i os
    | Just v <- val = jump f (n + 1) (i + v) (os & ix i +~ f v)
    | otherwise = n
    where val = os ^? ix i

part1 :: Tape -> Int
part1 = jump (const 1) 0 0

part2 :: Tape -> Int
part2 = jump f 0 0
    where f x = if x >= 3 then -1 else 1

main :: IO ()
main = do
    offsets <- parse <$> readFile "day5/input.txt"
    print . part1 $ offsets
    print . part2 $ offsets